File: lib/data/model/movie_model.dart
import 'dart:convert';

import '../../domain/model/movie_item.dart';

class MovieModel extends Movie {
  MovieModel({
    required super.id,
    required super.title,
    required super.overview,
    required super.posterPath,
    required super.backdropPath,
    required super.releaseDate,
    required super.runtime,
    required super.voteAverage,
    required super.tagline,
    required super.genres,
    required super.adult,
    required super.originalLanguage,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'overview': overview,
      'poster_path': posterPath,
      'backdrop_path': backdropPath,
      'release_date': releaseDate,
      'runtime': runtime,
      'vote_average': voteAverage,
      'tagline': tagline,
      'genres': jsonEncode(genres),
      'adult': adult ? 1 : 0,
      'original_language': originalLanguage,
      'updated_at': DateTime.now().millisecondsSinceEpoch,
    };
  }


  factory MovieModel.fromJson(Map<String, dynamic> json) {
    final List<dynamic>? genreDecoded = jsonDecode(json['genres']);

    return MovieModel(
      id: json['id'],
      title: json['title'],
      overview: json['overview'],
      posterPath: json['poster_path'] ?? '',
      backdropPath: json['backdrop_path'] ?? '',
      releaseDate: json['release_date'] ?? '',
      runtime: json['runtime'] ?? 0,
      voteAverage: (json['vote_average'] as num).toDouble(),
      tagline: json['tagline'] ?? '',
      genres: genreDecoded?.map((e) => e as String).toList() ?? [],
      adult: (json['adult'] as int) == 1,
      originalLanguage: json['original_language'],
    );
  }

  factory MovieModel.fromNetwork(Map<String, dynamic> json) {
    return MovieModel(
        id: json['id'],
        title: json['title'],
        overview: json['overview'],
        posterPath: json['poster_path'] ?? '',
        backdropPath: json['backdrop_path'] ?? '',
        releaseDate: json['release_date'] ?? '',
        runtime: json['runtime'] ?? 0,
        voteAverage: (json['vote_average'] as num).toDouble(),
        tagline: json['tagline'] ?? '',
        genres:
        (json['genres'] as List?)?.map((e) => e['name'] as String).toList() ??
            [],
        adult: (json['adult'] as bool),
        originalLanguage: json['original_language']
    );
  }
}

File: android/app/src/main/kotlin/com/gayuh/movies/core/remote/dto/MovieModel.kt
package com.gayuh.movies.core.remote.dto

import com.gayuh.movies.core.database.entity.MovieEntity
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class MovieModel(
    @Json(name = "id")
    val id: Int,

    @Json(name = "title")
    val title: String = "",

    @Json(name = "overview")
    val overview: String = "",

    @Json(name = "poster_path")
    val posterPath: String = "",

    @Json(name = "backdrop_path")
    val backdropPath: String = "",

    @Json(name = "release_date")
    val releaseDate: String = "",

    @Json(name = "runtime")
    val runtime: Int = 0,

    @Json(name = "vote_average")
    val voteAverage: Double = 0.0,

    @Json(name = "tagline")
    val tagline: String = "",

    @Json(name = "genres")
    val genres: List<GenreDto> = emptyList(),

    @Json(name = "adult")
    val adult: Boolean = false,

    @Json(name = "original_language")
    val originalLanguage: String = ""
) {
    companion object {
        fun MovieModel.toEntity(): MovieEntity =
            MovieEntity(
                id = id,
                title = title,
                overview = overview,
                posterPath = posterPath,
                backdropPath = backdropPath,
                releaseDate = releaseDate,
                runtime = runtime,
                voteAverage = voteAverage,
                tagline = tagline,
                genres = genres.joinToString(
                    prefix = "[",
                    postfix = "]"
                ) { "\"${it.name}\"" },
                adult = if (adult) 1 else 0,
                originalLanguage = originalLanguage,
                updatedAt = System.currentTimeMillis()
            )
    }
}
